### DISCUSSION TITLE
New Setting for `$PATH`

### DISCUSSION THREAD

**Discussion by chrisgrieser** [Taitava/obsidian-shellcommands#150] [Ideas - done & released]
Thanks to nil's fix, I finally got the plugin working to try out some stuff.

So I am trying to use markdownlint on the file, and everything works fine, but I do have to add stuff to `PATH` for basically every command that runs something that is installed, which is not only inconvenient, but also decreases the readability of your commands.

```sh
export PATH=/usr/local/bin/:/opt/homebrew/bin/:$PATH ; markdownlint --fix {{file_path:absolute}}
```

I propose to have a setting which paths to add to the PATH? 🙂

**Comment by Taitava** [comment]
Nice idea! 🙂

I think I would like to combine it with a more generic idea of implementing support for defining freely nameable environment variables: #23 (that discussion contains a bit old information about escaping variable values, but I think it has some pretty solid basic ideas).

Two things that would be needed so that it would work for your use-case:
 - An ability to define environment variables that are used for all shell commands, not just on a per shell command basis, as designed in the abovementioned discussion.
 - An ability to use the current value of the environment variable as part of the value. I.e. when defining `$PATH`, the purpose is not to overwrite it completely, but to append content there. Just like you have in your example:
> `export PATH=/usr/local/bin/:/opt/homebrew/bin/:$PATH`

Maybe the syntax for defining an environment variable in SC could be about something like this:
`/usr/local/bin/:/opt/homebrew/bin/:{{environment:PATH}}` (this would be written in a UI where you would first give the environment variable name `PATH`).

# Or just keep it simple?

Your original idea might be more simpler for a user to set up in the UI, as it would not require the user to think about including the old value of `$PATH` in the new value (this would be done automatically by SC).

I need to think about this and decide how to proceed.

Thank you! 🌞

**Reply by Taitava** [reply]
# Start with a dedicated setting for `$PATH`

(#23 can be implemented later).

Some decisions:

- Three multiline textareas (one for each operating system, Linux, macOS, Windows, to support sharing vaults/settings between multiple devices).
- Each line of a textarea will hold one directory. SC will concatenate all lines into a single line, using `:` or `;` as a separator. User can still put everything on one line if they want, they just need to use the correct separator themselves.
- The separator is determined automatically from which shell is used: `:` for Bash and the like, `;` for `PowerShell` and `CMD`.
- Support `{{variables}}`. I don't know if it's really needed, but one could e.g. include the current note's folder into `$PATH` by using `{{folder_path:absolute}}`, or define a `$PATH` directory in a note's YAML frontmatter (`{{yaml_value:custom_directory_to_add_to_path}}`).
- Create a new variable `{{environment:env_variable_name_here}}`, that can be used to retrieve the current value of `$PATH`.

## Add to the beginning or end of `$PATH`

Let's take a few examples of contents of an _Add to $PATH_ textarea:
```
/first/path
{{environment:path}}
/second/path
```

In this case, `/first/path` is added before the existing paths in `$PATH`. It's good when wanting to override default programs, but may have some negative performance effects.

`/second/path` is appended after `$PATH`. I think this is the recommended way to add stuff to `$PATH`, as it won't affect performance, and most of the time there's no need to override current programs.

---

```
{{environment:path}}
/first/path
/second/path
```
Everything is appended to the end of `$PATH`.

---

```
/first/path
/second/path
{{environment:path}}
```
Everything is appended before `$PATH`.

---

```
/first/path
/second/path
```

If the current `$PATH` value is not present in the textarea, SC could notice it and append the new values at the **end** of `$PATH`.

**Comment by Taitava** [comment]
@chrisgrieser ~~This just came to my mind, and I haven't tried it, but what happens if you put your `$PATH` additions to `~/.bashrc` (assuming you are using Bash? I'm not sure if you are, but anyway)? Will it have an effect when executing commands via SC? Or is `~/.bashrc` only used when opening up a normal terminal? This is interesting to me, and something I'll probably try myself sometime.~~

Edit 2022-04-30: Nevermind, it does not work.

**Comment by zcysxy** [comment]
+1 for this feature! Before it's fully supported, I think a simple solution for macOS is to add `/usr/local/bin/` and `/opt/homebrew/bin/` to the path on the plugin side. These are the brew installation path, and I think most macOS users use brew to manage command line tools. So this will cover most cases.

> This just came to my mind, and I haven't tried it, but what happens if you put your `$PATH` additions to `~/.bashrc` (assuming you are using Bash? I'm not sure if you are, but anyway)? Will it have an effect when executing commands via SC? Or is `~/.bashrc` only used when opening up a normal terminal? This is interesting to me, and something I'll probably try myself sometime.

This doesn't work for me. But we can do

```shell
export PATH="/usr/local/bin:/opt/homebrew/bin/:$PATH" && source ~/.zshrc # && do_something_else
```

The above command also solves the nvm/node problem here #191. Since the node path is added dynamically in my `~/.zhsrc`.

**Reply by Taitava** [reply]
> +1 for this feature! Before it's fully supported, I think a simple solution for macOS is to add `/usr/local/bin/` and `/opt/homebrew/bin/` to the path on the plugin side. These are the brew installation path, and I think most macOS users use brew to manage command line tools. So this will cover most cases.

The `$PATH` settings is actually already implemented (issue #204), but not released yet. I just haven't updated this conversation. I think I'll add a comment about it below with some screenshot. It will be released in `0.13.0`, so not in the next release (`0.12.0`) but after that. I'll probably make a beta test for `0.13.0` after the currently [on-going beta test for `0.12.0`](https://github.com/Taitava/obsidian-shellcommands/discussions/201) is finished.

> > This just came to my mind, and I haven't tried it, but what happens if you put your `$PATH` additions to `~/.bashrc` (assuming you are using Bash? I'm not sure if you are, but anyway)? Will it have an effect when executing commands via SC? Or is `~/.bashrc` only used when opening up a normal terminal? This is interesting to me, and something I'll probably try myself sometime.
> 
> This doesn't work for me.

Yea, I guess I noticed at some point it doesn't work 🙁. I guess [Node.js's `child_process`](https://nodejs.org/api/child_process.html) does not use the `.bashrc` file (or any similar file) when executing shell commands.

> But we can do
> 
> ```shell
> export PATH="/usr/local/bin:/opt/homebrew/bin/:$PATH" && source ~/.zshrc # && do_something_else
> ```
> 
> The above command also solves the nvm/node problem here #191. Since the node path is added dynamically in my `~/.zhsrc`.

It's a good workaround until `0.13.0` is out. 🙂👍

**Comment by Taitava** [comment]
# Implemented, waiting for a release OR a beta test (`0.13.0`)

The `$PATH` settings will be located in the _Environments_ tab (renamed from _Operating systems & shells_), below _Default shell_ settings.

![kuva](https://user-images.githubusercontent.com/13002142/166110238-7eb7889e-56ec-4d2f-862f-f594ebedf480.png)
This is a simple example adding just a couple of example directories mentioned by @zcysxy. There are three `$PATH` fields so that you can define different sets for macOS, Linux and Windows in case you happen to synchronise your Obsidian vault between different operating systems.

![kuva](https://user-images.githubusercontent.com/13002142/166110415-497e1538-fb9a-4743-8566-a59a2a3e680f.png)
This example shows how you can decide whether to insert the additional directories before or after the current content of `$PATH`. `{{!environment:PATH}}` gives you the current content of `$PATH`. If `{{!environment:PATH}}` is not present in the textarea, then the textarea's content will be appended **at the end of** `$PATH`.

Edit 2022-05-02: `{{!environment:PATH}}` should contain the `!` because special characters should **not** be escaped in `PATH`.

## See what `$PATH` currently contains
![kuva](https://user-images.githubusercontent.com/13002142/166110553-e913050a-7e48-4125-8017-1fceac981958.png)
In case you want to check what `$PATH` currently contains (without your additions), there's a (bit clumsy) modal that reveals it. The text lines are too long for the modal, but there's a horizontal scroll bar, so you should be able to see all of the content.

**Comment by Taitava** [comment]
# Beta test

https://github.com/Taitava/obsidian-shellcommands/discussions/228

**Comment by Taitava** [comment]
# Released
@chrisgrieser please let me know if you have any comments. 🙂
